source/frontend/src/features/properties/map/compensation/update/financials/FinancialActivitiesSubForm.tsx

```diff
@@ -0,0 +1,209 @@
+import { LinkButton, StyledRemoveLinkButton } from 'components/common/buttons';
+import { FastCurrencyInput, Select, SelectOption } from 'components/common/form';
+import GenericModal from 'components/common/GenericModal';
+import { SectionField } from 'features/mapSideBar/tabs/SectionField';
+import { FieldArray, FormikProps, useFormikContext } from 'formik';
+import { useState } from 'react';
+import { FaTrash } from 'react-icons/fa';
+import styled from 'styled-components';
+import { stringToBoolean } from 'utils/formUtils';
+
+import { CompensationRequisitionFormModel, FinacialActivityFormModel } from '../../models';
+
+export interface IFinancialActivitiesSubFormProps {
+  formikProps: FormikProps<CompensationRequisitionFormModel>;
+  compensationRequisitionId: number;
+  financialActivityOptions: SelectOption[];
+  gstConstant: number;
+  activitiesUpdated: () => void;
+}
+
+export const FinancialActivitiesSubForm: React.FunctionComponent<
+  IFinancialActivitiesSubFormProps
+> = ({
+  formikProps,
+  compensationRequisitionId,
+  financialActivityOptions,
+  gstConstant,
+  activitiesUpdated,
+}) => {
+  const { values, setFieldValue } = useFormikContext<CompensationRequisitionFormModel>();
+  const [showModal, setShowModal] = useState(false);
+  const [rowToDelete, setRowToDelete] = useState<number | undefined>(undefined);
+
+  const onUpdateGstApplicable = (index: number, gstOption: string): void => {
+    const isGstRequired = stringToBoolean(gstOption);
+    setAmountFields(index, isGstRequired, values.financials[index].pretaxAmount);
+  };
+
+  const onPretaxAmountUpdated = (index: number, newValue: string): void => {
+    const isGstRequired = stringToBoolean(values.financials[index].isGstRequired);
+    const cleanValue = getCurrenclyCleanValue(newValue);
+
+    setAmountFields(index, isGstRequired, cleanValue);
+  };
+
+  const setAmountFields = (index: number, gstRequired: boolean, pretaxAmount: number): void => {
+    let totalAmount = 0;
+    let taxAmount = 0;
+
+    if (gstRequired) {
+      taxAmount = pretaxAmount * gstConstant;
+      totalAmount = taxAmount + pretaxAmount;
+    } else {
+      totalAmount = pretaxAmount;
+    }
+
+    setFieldValue(`financials[${index}].taxAmount`, taxAmount);
+    setFieldValue(`financials[${index}].taxAmount`, taxAmount);
+    setFieldValue(`financials[${index}].totalAmount`, totalAmount);
+
+    activitiesUpdated();
+  };
+
+  return (
+    <FieldArray
+      name="financials"
+      render={arrayHelpers => {
+        return (
+          <>
+            {values.financials.map((financial, index) => (
+              <div key={`financial-act-${index}`}>
```

Comment:
in general we want to avoid using the index to form the key, as per: 

You might be tempted to use an item’s index in the array as its key. In fact, that’s what React will use if you don’t specify a key at all. But the order in which you render items will change over time if an item is inserted, deleted, or if the array gets reordered. Index as a key often leads to subtle and confusing bugs.

Similarly, do not generate keys on the fly, e.g. with key={Math.random()}. This will cause keys to never match up between renders, leading to all your components and DOM being recreated every time. Not only is this slow, but it will also lose any user input inside the list items. Instead, use a stable ID based on the data.

Note that your components won’t receive key as a prop. It’s only used as a hint by React itself. If your component needs an ID, you have to pass it as a separate prop: <Profile key={id} userId={id} />.

Reply:
updated